### PR TITLE
Rhai shortcodes: fixing performance issues

### PR DESCRIPTION
**Pull request by karthik2804**
Adapting #56 to include Rhai shortcodes support for markdown content. Adds a new page head field `enable_shortcodes`. It does not introduce a significant performance hit as it only runs the markdown content with the shortcode switch through handlebars before the markdown parser.

### CODE REVIEW

File: docs/content/shortcodes.md
Comment (vdice) [inline docs/content/shortcodes.md]:
```suggestion
Bartholomew supports shortcodes to simplify sharing content between different markdown files.
```

File: docs/content/shortcodes.md
Comment (vdice) [inline docs/content/shortcodes.md]:
```suggestion
An example of a shortcode is in `shortcodes/alert.rhai`. The shortcode is available in the mardown file as `alert` just like for scripts.
```

File: docs/content/shortcodes.md
Comment (vdice) [inline docs/content/shortcodes.md]:
```suggestion
To enable shortcode support for a particular document, the value of `enable_shortcodes = true` must be set in the page head.
```

File: docs/content/shortcodes.md
Comment (vdice) [inline docs/content/shortcodes.md]:
```suggestion
Now the shortcode can be used in the markdown file by calling it with the required arguments. For the alerts script, this is the type of alert and the message to be displayed.
```

File: docs/content/shortcodes.md
Comment (vdice) [inline docs/content/shortcodes.md]:
```suggestion
An important note to be considered while using shortcodes is that the `\{{ }}` must be escaped if they are not meant to be executed. This is only required in the content files where shortcodes are enabled.
```

File: docs/content/scripting.md
Comment (vdice) [inline docs/content/scripting.md]: This should stay as `{{ echo "world" }}` so that the rhai script actually executes, right?

File: docs/content/shortcodes.md
Comment (vdice) [inline docs/content/shortcodes.md:92]: In my testing, with `enable_shortcodes = true`, using `\{{ alert "warning" "Bartholomew is a work in progress" }}` results in:
Screen Shot 2022-07-29 at 10 40 12 AM

Whereas, using `{{ alert "warning" "Bartholomew is a work in progress" }}` results in expected behavior:

Screen Shot 2022-07-29 at 10 38 49 AM

Can you double-check?

File: docs/content/shortcodes.md
Comment (vdice) [inline docs/content/shortcodes.md]: See comment above about using the unescaped shortcode which seems to work properly.  If so, perhaps this should just be `\{{ alert "warning" "Bartholomew is a work in progress" }}` when users don't want shortcodes to execute?

**Review by vdice — COMMENTED** [review]
Some docs suggestions and then questions around escaping vs not escaping shortcodes.

Also, I'm not savvy on checking for performance as mentioned in the issue -- can @radu-matei advise/test?

File: docs/content/shortcodes.md
Comment (karthik2804) [reply docs/content/shortcodes.md:92]: This is the expected behavior. The `\` escapes the expression from being parsed by the rendering engine allowing it to just return it as a string when the users don't need the short code to execute.

File: docs/content/shortcodes.md
Comment (karthik2804) [reply docs/content/shortcodes.md]: The extra escaping is due to wanting to print the backslash as well. Using `\{{ alert "warning" "Bartholomew is a work in progress" }}` results in the rendering engine returning "{{ alert "warning" "Bartholomew is a work in progress" }}".  whereas using `\\{\{ alert "warning" "Bartholomew is a work in progress" }}` returns as "\\{{ alert "warning" "Bartholomew is a work in progress" }}".

File: docs/content/scripting.md
Comment (karthik2804) [reply docs/content/scripting.md:43]: The script is not expected to execute there as it demonstrates how the script can be called within a template.

Comment (karthik2804) [reply docs/content/scripting.md]: Yes as the shortcodes are not enabled for the scripting content. Also, the Rhai script isn't supposed to execute as it is just there to demonstrate how a script can be called form the template.

File: docs/content/shortcodes.md
Comment (vdice) [reply docs/content/shortcodes.md:92]: Ah I see, yup, it helps for me to look at the rendered documentation :)

File: docs/content/scripting.md
Comment (vdice) [reply docs/content/scripting.md]: I think since this scripting markdown doesn't have `enable_shortcodes = true`, we should leave this example unescaped.  Because currently the rendered docs show this:

Screen Shot 2022-07-29 at 11 03 35 AM

But don't we want it to show the following?

Screen Shot 2022-07-29 at 11 04 21 AM

File: docs/content/shortcodes.md
Comment (vdice) [reply docs/content/shortcodes.md]: I see.  Tricky one.  The rendered docs show this:
Screen Shot 2022-07-29 at 11 05 21 AM

I'm worried users will be confused thinking they need to use `\\{\{ alert "warning" "Bartholomew is a work in progress" }}` to escape the shortcode when `\{{ alert "warning" "Bartholomew is a work in progress" }}` is the normal way.  

Maybe instead of working around the `enable_shortcodes = true` on this markdown, we can take screenshots?  Not sure.

Comment (karthik2804) [reply docs/content/shortcodes.md]: I will revisit it soon and see how to fix it. Thanks for pointing it out.

File: docs/content/scripting.md
Comment (karthik2804) [reply docs/content/scripting.md]: Yes that is correct, I will make the change.

File: docs/content/shortcodes.md
Comment (karthik2804) [reply docs/content/shortcodes.md]: Fixed using the `{{{raw}}} ..... {{{{/raw}}}` tag.

Comment (vdice) [reply docs/content/shortcodes.md]: 🙌 Nice find!

**Review by vdice — APPROVED** [review]
Docs and functionality LGTM!

**Comment by vdice** [comment]
I ran similar tests as mentioned in https://github.com/fermyon/bartholomew/pull/56#issuecomment-1088645883 against the Bartholomew docs site and noticed perhaps a slight decrease in performance, but I wouldn't say enough to block this feature.  Thoughts @radu-matei ?

```
# with bartholomew.wasm built from main
$ bombardier -c 1 -n 1000 http://localhost:3000/configuration
Bombarding http://localhost:3000/configuration with 1000 request(s) using 1 connection(s)
 1000 / 1000 [===========================================================================================================================================================================================================================================] 100.00% 262/s 3s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec       270.62      32.70     304.91
  Latency        3.70ms   575.93us    19.79ms
  HTTP codes:
    1xx - 0, 2xx - 1000, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     6.01MB/s

# with bartholomew.wasm built from this branch
$ bombardier -c 1 -n 1000 http://localhost:3000/configuration
Bombarding http://localhost:3000/configuration with 1000 request(s) using 1 connection(s)
 1000 / 1000 [===========================================================================================================================================================================================================================================] 100.00% 226/s 4s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec       230.01      39.73     593.35
  Latency        4.38ms   684.70us    23.39ms
  HTTP codes:
    1xx - 0, 2xx - 1000, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     5.12MB/s
```

**Comment by karthik2804** [comment]
These results are pretty close to what I have.  There is some contribution to the decrease in performance due to this branch having an extra page in `shortcodes.md` to render.

A catch with this is that the more pages that use shortcodes, the slower the rendering time will be. It is because currently, we render every page on every request. Therefore the number of content pages we have will have an impact on the performance metrics.

File: docs/content/configuration.md
Comment (radu-matei) [inline docs/content/configuration.md]: *Rhai* scripts (capitalization)

File: docs/content/scripting.md
Comment (radu-matei) [inline docs/content/scripting.md]: Extra new line not needed here.

File: src/content.rs
Comment (radu-matei) [inline src/content.rs:60]: This should have a doc comment explaining what it is.

File: src/content.rs
Comment (radu-matei) [inline src/content.rs]: There seems to be some inconsistency between `shortcode` and `shortcodes` in various documentation and identifier names.

File: src/content.rs
Comment (radu-matei) [inline src/content.rs]: Not sure this comment is relevant here.

File: src/content.rs
Comment (radu-matei) [inline src/content.rs]: nit: would this statement be simplified with a `match` statement instead of an `if let`?

File: src/content.rs
Comment (radu-matei) [inline src/content.rs]: We really want to avoid calling `.unwrap()` on results, as this causes a panic.

Most likely, given that this function can now fail, is should probably return a `Result`, the call to `load_shortcode_dir` would propagate the error (`?`), and we should let the consumer decide how to handle the potential error.

**Review by radu-matei — COMMENTED** [review]
I have a few comments, but nothing major.

Re: performance, the fact that we need to render the additional shortcode does make sense, so I wouldn't be worried by the additional latency here — particularly because you're not incurring it unless you are actively using them.

See below:

```
$ bombardier -n 100 -c 1 http://localhost:3000/shortcodes
Bombarding http://localhost:3000/shortcodes with 100 request(s) using 1 connection(s)

Done!
Statistics        Avg      Stdev        Max
  Reqs/sec       193.15      33.53     244.26
  Latency        5.18ms     1.09ms    14.69ms
  HTTP codes:
    1xx - 0, 2xx - 100, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     4.96MB/s
$ bombardier -n 100 -c 1 http://localhost:3000/shortcodes
Bombarding http://localhost:3000/shortcodes with 100 request(s) using 1 connection(s)

Done!
Statistics        Avg      Stdev        Max
  Reqs/sec       247.07      48.25     305.95
  Latency        4.07ms     1.06ms    13.19ms
  HTTP codes:
    1xx - 0, 2xx - 100, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     6.02MB/s
```

Happy to do another pass at this once you address the comments.
Thanks!

File: docs/content/shortcodes.md
Comment (radu-matei) [inline docs/content/shortcodes.md]: typo: `mardown`

**Comment by radu-matei** [comment]
The last commit you pushed appears as unverified.
Would you mind keeping just a few useful commits and make sure they have a sign-off and GPG signature?

Thanks!

**Comment by karthik2804** [comment]
I am sorry about the last commit being unverified. There was a mess up on my part switching emails.  Would you recommend that I squash the commits in this PR?

**Comment by radu-matei** [comment]
Yeah, rebasing and squashing the unnecessary commits makes sense.

**Comment by vdice** [comment]
It looks like this is ready to go!  Defer to @radu-matei who had the most recent requested changes.

**Review by radu-matei — APPROVED** [review]
LGTM

**Comment by vdice** [comment]
Thank you for this great feature @karthik2804!

**Pull request merged**
